File: Lab5/svm.R
ptm = proc.time()
data.train = read.csv("train.csv",header = FALSE)   #reading train and test parts of dataset
data.test = read.csv("test.csv",header = FALSE)
print(proc.time() - ptm)

ptm <- proc.time()
svm.model = svm(V5~., data = data.train, kernel = "radial", gamma = 1, cachesize = 200) 
#V5 is a column with class labels. If class labels are in n-th column, then Vn is a column in R
prediction = predict(svm.model,data.test)

print(proc.time() - ptm)

#predictions can be expressed as double values - but at the should be divided in two classes
for(i in 1:length(prediction))
{
  if(prediction[i] > 0)
    prediction[i] = 1
  else
    prediction[i] = -1
}

table(prediction,data.test$V5)

binarization <- function(df) #function which binarizes any categorical column
{
  N = dim(df)[1]
  M = matrix(0,nrow = N,ncol = 0)
  for(i in 1:length(df))
  {
    if(is.factor(df[,i]))
    {
      n = length(levels(df[,i]))
      if(n == 2)
      {
        V = matrix(0,nrow = N,ncol = 1)
        for(j in 1:N)
        {
          f = levels(df[,i])[1]
          if(df[j,i] == f)
            V[j,1] = 1
          else
            V[j,1] = 0
        }
        M = cbind(M,V)
      }
      else
      {
        for(k in 1:n)
        {
          V = matrix(0,nrow = N,ncol = 1)
          for(j in 1:N)
          {
            f = levels(df[,i])[k]
            if(df[j,i] == f)
              V[j,1] = 1
            else
              V[j,1] = 0
          }
          M = cbind(M,V)
        }
      }
    }
    else
      M = cbind(M,df[,i])
  }
  return(M)
}

File: Lab5/svm.py
import io
import numpy as np
from sklearn import svm
import scipy as sp
import pandas as pd
import time

train = np.genfromtxt("", delimiter=',') #There should be names of files, which contain train and test parts of dataset
test = np.genfromtxt("", delimiter=',')
n = 102 #number of features (class labels not included)

classifier = svm.SVC(kernel='rbf', max_iter=200000, cache_size=8000)
t1 = time.time()
t2 = time.perf_counter()
t3 = time.process_time()

classifier.fit(train[:,0:n],train[:,n])

prediction = classifier.predict(test[:,0:n])

print("Total time = ", time.time() - t1)
print("Performance time = ", time.perf_counter() - t2)
print("Processor time = ", time.process_time() - t3)

print(np.sum(prediction))

tab = pd.crosstab(index=prediction, columns=test[:,n]) #confusion matrix

print(tab)
